datafusion/substrait/src/logical_plan/consumer.rs

```diff
@@ -1682,18 +1683,35 @@ pub async fn from_substrait_rex(
             Ok(Expr::Literal(scalar_value))
         }
         Some(RexType::Cast(cast)) => match cast.as_ref().r#type.as_ref() {
-            Some(output_type) => Ok(Expr::Cast(Cast::new(
-                Box::new(
-                    from_substrait_rex(
-                        state,
-                        cast.as_ref().input.as_ref().unwrap().as_ref(),
-                        input_schema,
-                        extensions,
-                    )
-                    .await?,
-                ),
-                from_substrait_type_without_names(output_type, extensions)?,
-            ))),
+            Some(output_type) => {
+                if cast.failure_behavior() == ReturnNull {
+                    Ok(Expr::TryCast(TryCast::new(
+                        Box::new(
+                            from_substrait_rex(
+                                state,
+                                cast.as_ref().input.as_ref().unwrap().as_ref(),
+                                input_schema,
+                                extensions,
+                            )
+                            .await?,
+                        ),
+                        from_substrait_type_without_names(output_type, extensions)?,
+                    )))
+                } else {
+                    Ok(Expr::Cast(Cast::new(
+                        Box::new(
+                            from_substrait_rex(
+                                state,
+                                cast.as_ref().input.as_ref().unwrap().as_ref(),
+                                input_schema,
+                                extensions,
+                            )
+                            .await?,
+                        ),
+                        from_substrait_type_without_names(output_type, extensions)?,
```

Comment:
minor: In both branches, the expression and data type we are creating and using is identical. It would be be good to pull their creating out of the if-else to avoid duplication.